### PR TITLE
New document: Choosing a framework

### PR DESCRIPTION
**Pull request by nottrobin**
This document lays out the reasons for choosing one application structure over another - e.g. Jekyll vs Flask, Flask vs Django.

I've also restructured the documents slightly to create a new section called "project structure".

Please read it and let me know what you think.

### CODE REVIEW

File: project-structure/choosing-a-framework.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+For dynamic server-side applications, our preferred language is [Python](https://www.python.org/), a mature language with a clear focus on ease-of-use and friendliness and extensive support for web development. Our expertise is in Python, but we remain open to exploring other languages if sensible opportunities arise.
+
+#### Flask for database-less applications
+[Flask](http://flask.pocoo.org/) is a simple microframework for web development with sensible defaults. For web applications that don't require a database back-end, Flask is our recommended framework as it gives you the flexibility to add as much or as little extra logic as needed.
+
+*Projects using Flask: [snapcraft.io](https://github.com/canonical-websites/snapcraft.io); [blog.ubuntu.com](https://github.com/canonical-websites/blog.ubuntu.com); [jaas.ai](https://github.com/canonical-websites/jaas.ai); [docs.snapcraft.io](https://github.com/canonical-websites/docs.ubuntu.com)*
```
Comment (Lukewh) [inline project-structure/choosing-a-framework.md]: [docs.snapcraft.io] (https://.../docs.ubuntu.com)

File: project-structure/choosing-a-framework.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
+# Choosing a framework for a web application
+An outline of the philosophy and rationale behind our recommendations for choosing software and structures for [our website applications](https://github.com/canonical-websites/).
+
+## Philosophy
+We aim to always encourage participation in our projects from both within the company and the general public. We aim to support and grow the communities that surround our work. We do this by creating web applications that are [simple](https://en.wikipedia.org/wiki/KISS_principle) for everyone to understand and use.
+
+Our work should be [available to the public](https://en.wikipedia.org/wiki/Open_by_default), and as easy as possible for the anyone to [contribute to](https://en.wikipedia.org/wiki/Open_collaboration) or learn from. This philosophy should permeate everything we do, from the software we choose to how we write [READMEs](https://en.wikipedia.org/wiki/README) and [our code](https://developer.gnome.org/programming-guidelines/stable/writing-good-code.html.en), to how we respond to pull requests and issues.
```
Comment (Lukewh) [inline project-structure/choosing-a-framework.md]: "for the anyone to [contribute to]"

**Comment by nottrobin** [comment]
Thanks @Lukewh, I've addressed those two.

**Comment by squidsoup** [comment]
Hey @nottrobin, I don't see anything in here about building apps in js. That seems like a significant omission.

**Comment by nottrobin** [comment]
@squidsoup yes, I haven't mentioned client-side software choices much, deliberately, I'm not the right person to write that part. I'd welcome a follow-up PR to add guidelines on client-side code after this one is merged, but I'd rather keep the scope limited to the topics I've already covered in this PR itself.

**Comment by nottrobin** [comment]
@hatched just BTW, approving PRs on this repo is a bit different than on most repos:

https://github.com/canonical-webteam/practices/blob/master/CONTRIBUTING.md#approval-criteria
vs
https://github.com/canonical-webteam/practices/blob/master/workflow/github.md#merging-process

File: project-structure/server-side-frameworks.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+layout: base
+---
+
+# Server-side framework suggestions
+
+These are our suggested server-side technologies to use for website applications in specific scenarios, which should always sit within the usual [general project structure](general-structure). The focus here is on keeping the application topology [as simple as possible](project-philosophy).
+
+## Static site generators
+
+For a basic website that doesn't require any dynamic processing on the server, using a [static site generator](https://davidwalsh.name/introduction-static-site-generators) is recommended. Flat HTML pages will be generated at build time, leading to a deployment payload that is simply a flat collection of HTML files to be served with a simple HTTP server - usually Nginx.
```
Comment (b-m-f) [inline project-structure/server-side-frameworks.md]: How about  adding `[Nginx](https://nginx.org/en/)`?

**Comment by jpmartinspt** [comment]
Looks good.

File: project-structure/server-side-frameworks.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+*Projects using Hugo: [usn.ubuntu.com](https://launchpad.net/usn.ubuntu.com)*
+
+## Dynamic Python applications
+
+When you need logical processing on the server, you'll need a dynamic application. Examples of websites that need dynamic processing:
+
+- Websites with a database back-end (e.g. a CMS like [partners.ubuntu.com](https://partners.ubuntu.com))
+- Websites that retrieve their core content from an API (e.g. [snapcraft.io](https://snapcraft.io), [blog.ubuntu.com](https://blog.ubuntu.com))
+
+Many websites may needs to pull in related content from APIs - e.g. to show a list of related blog posts within a page. If the API content isn't central to the core application, this is usually best achieved with a client-side solution using JavaScript. This can gracefully degrade if the API cannot be reached without delaying or breaking the rest of the application. In these cases, you may not need a dynamic application.
```
Comment (b-m-f) [inline project-structure/server-side-frameworks.md]: `needs` -> `need`?

**Comment by nottrobin** [comment]
@b-m-f I created https://github.com/canonical-webteam/practices/pull/107 for your suggestions.

**Pull request merged**